storage/imx6ull-flash/imx6ull-flash-test.c

```diff
@@ -444,13 +519,166 @@ void test_ecc(void)
+		err = flashdrv_read(dma, paddr, data, aux);
+		if (err < 0) {
+			printf("read() failed: %d\n", err);
+			break;
+		}
+
+		/* Check and correct bitflips (adapted _flashmtd_checkECC() implementation) */
+		for (i = 0; i < sizeof(aux->errors); i++) {
+			switch (aux->errors[i]) {
+				case flash_no_errors:
+				case flash_erased:
+					break;
+
+				case flash_uncorrectable:
+					/* BCH reports chunk as uncorrectable in case of an erased page with bitflips within that chunk area */
+					/* Check if that's the case by counting the bitflips with an assumption that the page is fully erased */
+
+					/* Read raw page in order to check full chunk - both data and its ECC area */
+					if (raw == NULL) {
+						/* Map raw page buffer */
+						raw = mmap(NULL, pagemapsz, PROT_READ | PROT_WRITE, MAP_UNCACHED, OID_CONTIGUOUS, 0);
+						if (raw == MAP_FAILED) {
+							raw = NULL;
+							err = -ENOMEM;
+							printf("mmap() failed: %d\n", err);
+							break;
+						}
+
+						/* Read raw page */
+						err = flashdrv_readraw(dma, paddr, raw, FLASHDRV_PAGESZ);
+						if (err < 0) {
+							printf("readraw() failed: %d\n", err);
+							break;
+						}
+					}
+
+					/* Metadata chunk */
+					if (i == 0) {
+						boffs = 0;
+						blen = mlen;
+					}
+					/* Data chunk */
+					else {
+						boffs = mlen + (i - 1) * dlen;
+						blen = dlen;
+					}
+
+					flips = flashdrv_checkErased(raw, boffs, blen);
+
+					/* Nothing to do if there're no bitflips */
+					if (flips == 0) {
+						break;
+					}
+
+					/* Handle metadata chunk bitflips */
+					if (i == 0) {
+						/* Too many metadata bitflips, return error */
+						if (flips > maxmetaecc) {
+							err = -EBADMSG;
+							printf("too many bitflips in metadata chunk: %d\n", err);
+							break;
+						}
+
+						/* Correct metadata chunk */
+						memset(aux, 0xff, flashinfo.oobsz);
+					}
+					/* Handle data chunk bitflips */
+					else {
+						/* Too many data bitflips, return error */
+						if (flips > maxdataecc) {
+							err = -EBADMSG;
+							printf("too many bitflips in data chunk %d: %d\n", i, err);
+							break;
+						}
+
+						/* Correct data chunk */
+						memset(data + (i - 1) * 512, 0xff, 512);
+					}
+
+					/* Chunk corrected, update max number of bitflips */
+					maxflips = max(flips, maxflips);
+					break;
+
+				default:
+					/* Chunk corrected by BCH, update max number of bitflips */
+					maxflips = max(aux->errors[i], maxflips);
+			}
+
+			if (err < 0) {
+				break;
+			}
+		}
+
+		if (err < 0) {
+			break;
+		}
+
+		/* Verify metadata */
+		for (i = 0; i < flashinfo.oobsz; i++) {
+			if (aux->metadata[i] != 0xff) {
+				err = -EBADMSG;
+				printf("failed to correct metadata, %d\n", err);
+				break;
+			}
+		}
+
+		/* Verify data */
+		for (i = 0; i < flashinfo.writesz; i++) {
+			if (data[i] != 0xff) {
+				err = -EBADMSG;
+				printf("failed to correct data, %d\n", err);
+				break;
+			}
+		}
+
+		if (err < 0) {
+			break;
+		}
```

Comment:
You check the same condition after exit from `do {...} while(0);` It might be removed here.

Reply:
I guess it's to skip success message printf

Reply:
Yes, in case of error we don't print the message.

Reply:
```suggestion
		if (err < 0) {
			printf("[%4u]: ECC test failed\n", paddr);
		}
		else {
		        printf("[%4u]: successfully corrected erased page data%s\n", paddr, (maxflips >= thresecc) ? ", page needs rewrite" : "");
		        printf("[%4u]: ECC successfully corrected page data\n", paddr);
		}
```

Reply:
@HBuczynski suggested that I could combine the messages inside and outside the `do {} while (0)` loop since they're printed in the same case - no need to duplicate the `err <  0` check. I've added this change.